controllers/snapshotenvironmentbinding_controller.go

```diff
@@ -0,0 +1,194 @@
+	ApplicationLabelName                   = "appstudio.application"
+	EnvironmentLabelName                   = "appstudio.environment"
+	ComponentLabelName                     = "appstudio.component"
+	linkedRemoteSecretsTargetFinalizerName = "spi.appstudio.redhat.com/remote-secrets"
+)
+
+var (
+	unableToFetchRemoteSecretsError = stderrors.New("unable to fetch the Remote Secrets list")
+	unableToUpdateRemoteSecret      = stderrors.New("unable to update the remote secret")
+)
+
+// SnapshotEnvironmentBindingReconciler reconciles a SnapshotEnvironmentBinding object
+type SnapshotEnvironmentBindingReconciler struct {
+	k8sClient     client.Client
+	Scheme        *runtime.Scheme
+	Configuration *opconfig.OperatorConfiguration
+	finalizers    finalizer.Finalizers
+}
+
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=snapshotenvironmentbindings,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=snapshotenvironmentbindings/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=snapshotenvironmentbindings/finalizers,verbs=update
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=environments,verbs=get;list;watch
+
+func (r *SnapshotEnvironmentBindingReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.finalizers = finalizer.NewFinalizers()
+	if err := r.finalizers.Register(linkedRemoteSecretsTargetFinalizerName, &linkedRemoteSecretsFinalizer{client: r.k8sClient}); err != nil {
+		return fmt.Errorf("failed to register the linked remote secret finalizer: %w", err)
+	}
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&appstudiov1alpha1.SnapshotEnvironmentBinding{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Complete(r)
+
+	if err != nil {
+		return fmt.Errorf("failed to configure the reconciler: %w", err)
+	}
+	return nil
+}
+
+func (r *SnapshotEnvironmentBindingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	lg := log.FromContext(ctx)
+	lg.V(logs.DebugLevel).Info("starting reconciliation")
+
+	var snapshotEnvBinding appstudiov1alpha1.SnapshotEnvironmentBinding
+	if err := r.k8sClient.Get(ctx, req.NamespacedName, &snapshotEnvBinding); err != nil {
+		if errors.IsNotFound(err) {
+			lg.V(logs.DebugLevel).Info("SnapshotEnvironmentBinding already gone from the cluster. skipping reconciliation")
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, fmt.Errorf("failed to get the SnapshotEnvironmentBinding: %w", err)
+	}
+
+	finalizationResult, err := r.finalizers.Finalize(ctx, &snapshotEnvBinding)
+	if err != nil {
+		// if the finalization fails, the finalizer stays in place, and so we don't want any repeated attempts until
+		// we get another reconciliation due to cluster state change
+		return ctrl.Result{Requeue: false}, fmt.Errorf("failed to finalize: %w", err)
+	}
+	if finalizationResult.Updated {
+		if err = r.k8sClient.Update(ctx, &snapshotEnvBinding); err != nil {
```

Comment:
Are we going to update the SED ?

Reply:
yes it must set finalizer name into metadata